{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "preprocess_mesa.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "authorship_tag": "ABX9TyOmP99umEHK4hI8dbhFm9t4",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/collinwa/MPCCA/blob/main/analysis_notebooks/preprocess_mesa.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "PMnUxdiaDqnb"
      },
      "source": [
        "import glob\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import torch\n",
        "np.set_printoptions(linewidth = 200)"
      ],
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "H3pXdTn7TZHO"
      },
      "source": [
        "def filter_data(X, min_mean = None, min_var = None, max_missing = None):\n",
        "  drop = []\n",
        "  if min_mean is not None:\n",
        "    mean_X = X.mean()\n",
        "    drop += list(mean_X.index[mean_X <= min_mean])\n",
        "  if min_var is not None:\n",
        "    var_X = X.var()\n",
        "    drop += list(var_X.index[var_X <= min_var])\n",
        "  if max_missing is not None:\n",
        "    missing_X = pd.isna(X).mean()\n",
        "    drop += list(missing_X.index[missing_X > max_missing])\n",
        "  return X.drop(columns = drop)"
      ],
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "RM4ebPUQB1tQ"
      },
      "source": [
        "exam = '1'\n",
        "# exam = '5'\n",
        "\n",
        "# Basic datasets\n",
        "exp_fn = '/gpfs/commons/datasets/controlled/MESA/topmed_mesa_exchange/rna/phenotype_matrices/PBMC_Exam_' + exam + '.expression.bed.gz'\n",
        "methyl_fn = '/gpfs/commons/projects/MESA/projects/mapping_mqtls/input/data/topmed_mesa.exam_' + exam + '.multiethnic.normalized_methylation.bed.gz'\n",
        "metab_fn = '/gpfs/commons/datasets/controlled/MESA/topmed_mesa_exchange/metabolites/phenotype_matrices/Metabolites_Exam_' + exam + '.invnorm.parquet'\n",
        "protein_fn = '/gpfs/commons/datasets/controlled/MESA/topmed_mesa_exchange/proteins/phenotype_matrices/Proteins_Exam_' + exam + '.invnorm.parquet'\n",
        "\n",
        "# Probe info\n",
        "probe_snp_overlap_fn = '/gpfs/commons/projects/MESA/projects/annotation/EPIC_polymorphicprobes/freeze8.pass_only.phased.mesa_1319samples.maf01.biallelic.SNVs_overlapping_with_probes.txt.gz'\n",
        "probe_manifest_fn = '/gpfs/commons/projects/MESA/projects/annotation/EPIC_hg38/EPIC.hg38.manifest.tsv.gz'\n",
        "\n",
        "# Sample/phenotype/batch info (without exam labels?)\n",
        "sample_info_fn = '/gpfs/commons/projects/MESA/projects/summary_of_data/unique_individuals_across_molecular_data.txt'\n",
        "methyl_celltype_fn = '/gpfs/commons/projects/MESA/projects/preprocess_methylation/houseman/estimated_cellcounts_houseman.txt'\n",
        "exp_celltype_fn = '/gpfs/commons/datasets/controlled/MESA/topmed_mesa_exchange/deconvolution/TOPMed_MESA_RNAseq.2648_freeze_samples.cibersort_lm22.collapsed.txt'\n",
        "methyl_batch_fn = '/gpfs/commons/projects/MESA/projects/preprocess_methylation/samplesheet/samplesheet_with_feno.mixup_fix.only_BIS.pass_QC.no_duplicates.no_relatives.freeze_8.multiethnic.txt'\n",
        "exp_batch_fn = '/gpfs/commons/datasets/controlled/MESA/topmed_mesa_exchange/metadata/TOPMed_MESA_RNAseq_2973samples_metadata.txt'\n",
        "\n",
        "# Sample/phenotype/batch info with exam labels\n",
        "metab_batch_fn = '/gpfs/commons/datasets/controlled/MESA/topmed_mesa_exchange/metabolites/covariates/Metabolites_Exam_' + exam + '.txt'\n",
        "phenotype_fp = '/gpfs/commons/projects/MESA/projects/pheno_data/exam_' + exam + '.*.txt'"
      ],
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "vMdLwO5KCekT"
      },
      "source": [
        "methyl_data_save_fn='/gpfs/commons/projects/MESA/projects/mpcca/methyl_cca_collated_exam' + exam + '.csv'\n",
        "metab_data_save_fn='/gpfs/commons/projects/MESA/projects/mpcca/metab_cca_collated_exam' + exam + '.csv'\n",
        "exp_data_save_fn='/gpfs/commons/projects/MESA/projects/mpcca/exp_cca_collated_exam' + exam + '.csv'\n",
        "protein_data_save_fn='/gpfs/commons/projects/MESA/projects/mpcca/protein_cca_collated_exam' + exam + '.csv'\n",
        "meta_df_save_fn='/gpfs/commons/projects/MESA/projects/mpcca/metadata_cca_collated_exam' + exam + '.csv'"
      ],
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "XfU98ijKEVT1"
      },
      "source": [
        "sample_info = pd.read_table(sample_info_fn)\n",
        "metab_batch = pd.read_table(metab_batch_fn, index_col=0).T\n",
        "exp_batch = pd.read_table(exp_batch_fn, index_col=0)\n",
        "methyl_batch = pd.read_table(methyl_batch_fn, index_col='TOEID')\n",
        "\n",
        "methyl_ct = pd.read_table(methyl_celltype_fn, index_col='TOEID')\n",
        "exp_ct = pd.read_table(exp_celltype_fn, index_col=0)"
      ],
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "YDQsnpK-WI2z"
      },
      "source": [
        "# Note:There are some expression samples that are not in this sample_info table,\n",
        "#   so we can't rely on that to find our complete data samples and we need to\n",
        "#   manually intersect the data types.\n",
        "\n",
        "# Some columns contain multiple sample labels. Split them so there is one row per sample.\n",
        "split_columns = ['Metabolomics', 'Methylomics', 'Proteomics', 'Transcriptomics_PBMC']\n",
        "for column in split_columns:\n",
        "  sample_info[column] = sample_info[column].str.split(';')\n",
        "  sample_info = sample_info.explode(column)\n",
        "\n",
        "sample_info = sample_info[sample_info['Exam'] == int(exam)].drop(\n",
        "    columns=['Exam', 'Transcriptomics_Mono', 'Transcriptomics_Tcell',\n",
        "             'nr_molecular_file']).drop_duplicates()"
      ],
      "execution_count": 6,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "GiezIMew8w6I"
      },
      "source": [
        "methyl_batch = methyl_batch.drop(\n",
        "    columns=['Sample_Name', 'Sample_Group', 'Pool_ID', 'sidno', 'DecodedID',\n",
        "             'LEVELE1.id', 'exam', 'age', 'race', 'race.1', 'sex', 'sex.1',\n",
        "             'Basename', 'sidno_exam.ID', 'NWDid'])"
      ],
      "execution_count": 7,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "z1S2X_WACfSm"
      },
      "source": [
        "exp_batch = exp_batch.drop(\n",
        "    columns=['age', 'exam', 'latest_run_ended', 'participant_id', 'race',\n",
        "             'sample_type', 'sex', 'study', 'cohort', 'cohort2', 'has_genotype',\n",
        "             'phase', 'source', 'nwd_id'])"
      ],
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pheno_df = pd.concat(\n",
        "    [pd.read_table(fn, index_col=0) for fn in glob.glob(phenotype_fp)], axis=1).drop(\n",
        "    columns=['sex', 'sidno', 'age', 'race', 'race_1', 'sex_1', 'race_white',\n",
        "             'race_black', 'race_chinese', 'race_hispanic'])"
      ],
      "metadata": {
        "id": "Nfrv6CfYNMtV"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "lYIPGUkF3l0C"
      },
      "source": [
        "# exp_batch seems to include the samples that have extant RNA-seq data but NA\n",
        "#   for the sample id in the sample_info file.\n",
        "# meta_batch has an entry for every data point.\n",
        "# methyl_batch has an entry for every data point.\n",
        "meta_df = sample_info.join(\n",
        "    exp_batch, on='Transcriptomics_PBMC').join(\n",
        "    metab_batch[['column_2', 'column_3']], on='NWDID').join(\n",
        "    methyl_batch, on='Methylomics').join(\n",
        "    methyl_ct.drop(columns='IID'), on='Methylomics').join(\n",
        "    exp_ct, on='Transcriptomics_PBMC').join(\n",
        "    pheno_df, on='NWDID')\n",
        "meta_df['num_nans'] = pd.isna(meta_df).sum(1).values\n",
        "meta_df.index = meta_df.NWDID"
      ],
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "meta_df"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 475
        },
        "id": "DZ1Hb6zQsTbb",
        "outputId": "cc54c13d-61be-4b19-da00-25c121ff2acb"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "           sidno      NWDID     Sex      Race  Age Metabolomics Methylomics  \\\n",
              "NWDID                                                                         \n",
              "NWD133576  10008  NWD133576    male     white   61    TOM916213   TOE755769   \n",
              "NWD678628  10016  NWD678628  female  hispanic   74    TOM704590   TOE814141   \n",
              "NWD302101  10020  NWD302101  female     white   78    TOM657913   TOE814238   \n",
              "NWD852989  10024  NWD852989    male   chinese   52          NaN         NaN   \n",
              "NWD132273  10028  NWD132273    male     white   72    TOM645842   TOE491965   \n",
              "...          ...        ...     ...       ...  ...          ...         ...   \n",
              "NWD717781  26443  NWD717781    male     black   63          NaN         NaN   \n",
              "NWD109059  26444  NWD109059    male     black   73          NaN         NaN   \n",
              "NWD996654  26456  NWD996654  female     black   62          NaN         NaN   \n",
              "NWD164446  26461  NWD164446  female     black   54    TOM189936   TOE904921   \n",
              "NWD281909  26465  NWD281909    male     black   66    TOM434320   TOE901400   \n",
              "\n",
              "          Proteomics Transcriptomics_PBMC  3p_bias  ...        bmi  waist_cm  \\\n",
              "NWDID                                               ...                        \n",
              "NWD133576  TOP378136            TOR905983      NaN  ...  24.989422      96.0   \n",
              "NWD678628  TOP543330                  NaN      NaN  ...  30.228374     105.0   \n",
              "NWD302101  TOP464569            TOR498124     0.29  ...  20.835997      84.5   \n",
              "NWD852989        NaN            TOR818278      NaN  ...  22.739382      89.7   \n",
              "NWD132273  TOP608397            TOR191728     0.25  ...        NaN       NaN   \n",
              "...              ...                  ...      ...  ...        ...       ...   \n",
              "NWD717781        NaN            TOR365409      NaN  ...  24.106866      93.0   \n",
              "NWD109059        NaN            TOR264293     0.24  ...  34.540082     124.0   \n",
              "NWD996654        NaN            TOR690669      NaN  ...  28.348906      95.0   \n",
              "NWD164446  TOP744562            TOR701404     0.28  ...  37.513235     107.5   \n",
              "NWD281909  TOP648292                  NaN      NaN  ...  21.815754      88.6   \n",
              "\n",
              "          hip_cm moderate moderate_met moderate_total_met vigorous  \\\n",
              "NWDID                                                                \n",
              "NWD133576  106.0    120.0        660.0             1080.0      0.0   \n",
              "NWD678628  108.5      0.0          0.0             6930.0      0.0   \n",
              "NWD302101   93.5      0.0          0.0             6630.0      0.0   \n",
              "NWD852989   95.9      0.0          0.0             3030.0      0.0   \n",
              "NWD132273    NaN      NaN          NaN                NaN      NaN   \n",
              "...          ...      ...          ...                ...      ...   \n",
              "NWD717781  100.0    180.0        990.0             4860.0    120.0   \n",
              "NWD109059  114.2      0.0          0.0             1877.5      0.0   \n",
              "NWD996654  104.1    360.0       1980.0             6975.0      0.0   \n",
              "NWD164446  127.1      0.0          0.0              810.0      0.0   \n",
              "NWD281909   98.3      0.0          0.0             1560.0      0.0   \n",
              "\n",
              "          vigorous_met  vigorous_total_met  num_nans  \n",
              "NWDID                                                 \n",
              "NWD133576          0.0                 0.0        33  \n",
              "NWD678628          0.0                 0.0        30  \n",
              "NWD302101          0.0               390.0        15  \n",
              "NWD852989          0.0                 0.0        40  \n",
              "NWD132273          NaN                 NaN        87  \n",
              "...                ...                 ...       ...  \n",
              "NWD717781        840.0               840.0        34  \n",
              "NWD109059          0.0                 0.0        31  \n",
              "NWD996654          0.0                 0.0        36  \n",
              "NWD164446          0.0                 0.0        15  \n",
              "NWD281909          0.0                 0.0        30  \n",
              "\n",
              "[1438 rows x 122 columns]"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sidno</th>\n",
              "      <th>NWDID</th>\n",
              "      <th>Sex</th>\n",
              "      <th>Race</th>\n",
              "      <th>Age</th>\n",
              "      <th>Metabolomics</th>\n",
              "      <th>Methylomics</th>\n",
              "      <th>Proteomics</th>\n",
              "      <th>Transcriptomics_PBMC</th>\n",
              "      <th>3p_bias</th>\n",
              "      <th>...</th>\n",
              "      <th>bmi</th>\n",
              "      <th>waist_cm</th>\n",
              "      <th>hip_cm</th>\n",
              "      <th>moderate</th>\n",
              "      <th>moderate_met</th>\n",
              "      <th>moderate_total_met</th>\n",
              "      <th>vigorous</th>\n",
              "      <th>vigorous_met</th>\n",
              "      <th>vigorous_total_met</th>\n",
              "      <th>num_nans</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>NWDID</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>NWD133576</th>\n",
              "      <td>10008</td>\n",
              "      <td>NWD133576</td>\n",
              "      <td>male</td>\n",
              "      <td>white</td>\n",
              "      <td>61</td>\n",
              "      <td>TOM916213</td>\n",
              "      <td>TOE755769</td>\n",
              "      <td>TOP378136</td>\n",
              "      <td>TOR905983</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>24.989422</td>\n",
              "      <td>96.0</td>\n",
              "      <td>106.0</td>\n",
              "      <td>120.0</td>\n",
              "      <td>660.0</td>\n",
              "      <td>1080.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>33</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>NWD678628</th>\n",
              "      <td>10016</td>\n",
              "      <td>NWD678628</td>\n",
              "      <td>female</td>\n",
              "      <td>hispanic</td>\n",
              "      <td>74</td>\n",
              "      <td>TOM704590</td>\n",
              "      <td>TOE814141</td>\n",
              "      <td>TOP543330</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>30.228374</td>\n",
              "      <td>105.0</td>\n",
              "      <td>108.5</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>6930.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>30</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>NWD302101</th>\n",
              "      <td>10020</td>\n",
              "      <td>NWD302101</td>\n",
              "      <td>female</td>\n",
              "      <td>white</td>\n",
              "      <td>78</td>\n",
              "      <td>TOM657913</td>\n",
              "      <td>TOE814238</td>\n",
              "      <td>TOP464569</td>\n",
              "      <td>TOR498124</td>\n",
              "      <td>0.29</td>\n",
              "      <td>...</td>\n",
              "      <td>20.835997</td>\n",
              "      <td>84.5</td>\n",
              "      <td>93.5</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>6630.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>390.0</td>\n",
              "      <td>15</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>NWD852989</th>\n",
              "      <td>10024</td>\n",
              "      <td>NWD852989</td>\n",
              "      <td>male</td>\n",
              "      <td>chinese</td>\n",
              "      <td>52</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TOR818278</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>22.739382</td>\n",
              "      <td>89.7</td>\n",
              "      <td>95.9</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>3030.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>40</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>NWD132273</th>\n",
              "      <td>10028</td>\n",
              "      <td>NWD132273</td>\n",
              "      <td>male</td>\n",
              "      <td>white</td>\n",
              "      <td>72</td>\n",
              "      <td>TOM645842</td>\n",
              "      <td>TOE491965</td>\n",
              "      <td>TOP608397</td>\n",
              "      <td>TOR191728</td>\n",
              "      <td>0.25</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>87</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>NWD717781</th>\n",
              "      <td>26443</td>\n",
              "      <td>NWD717781</td>\n",
              "      <td>male</td>\n",
              "      <td>black</td>\n",
              "      <td>63</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TOR365409</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>24.106866</td>\n",
              "      <td>93.0</td>\n",
              "      <td>100.0</td>\n",
              "      <td>180.0</td>\n",
              "      <td>990.0</td>\n",
              "      <td>4860.0</td>\n",
              "      <td>120.0</td>\n",
              "      <td>840.0</td>\n",
              "      <td>840.0</td>\n",
              "      <td>34</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>NWD109059</th>\n",
              "      <td>26444</td>\n",
              "      <td>NWD109059</td>\n",
              "      <td>male</td>\n",
              "      <td>black</td>\n",
              "      <td>73</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TOR264293</td>\n",
              "      <td>0.24</td>\n",
              "      <td>...</td>\n",
              "      <td>34.540082</td>\n",
              "      <td>124.0</td>\n",
              "      <td>114.2</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1877.5</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>31</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>NWD996654</th>\n",
              "      <td>26456</td>\n",
              "      <td>NWD996654</td>\n",
              "      <td>female</td>\n",
              "      <td>black</td>\n",
              "      <td>62</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TOR690669</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>28.348906</td>\n",
              "      <td>95.0</td>\n",
              "      <td>104.1</td>\n",
              "      <td>360.0</td>\n",
              "      <td>1980.0</td>\n",
              "      <td>6975.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>36</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>NWD164446</th>\n",
              "      <td>26461</td>\n",
              "      <td>NWD164446</td>\n",
              "      <td>female</td>\n",
              "      <td>black</td>\n",
              "      <td>54</td>\n",
              "      <td>TOM189936</td>\n",
              "      <td>TOE904921</td>\n",
              "      <td>TOP744562</td>\n",
              "      <td>TOR701404</td>\n",
              "      <td>0.28</td>\n",
              "      <td>...</td>\n",
              "      <td>37.513235</td>\n",
              "      <td>107.5</td>\n",
              "      <td>127.1</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>810.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>15</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>NWD281909</th>\n",
              "      <td>26465</td>\n",
              "      <td>NWD281909</td>\n",
              "      <td>male</td>\n",
              "      <td>black</td>\n",
              "      <td>66</td>\n",
              "      <td>TOM434320</td>\n",
              "      <td>TOE901400</td>\n",
              "      <td>TOP648292</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>21.815754</td>\n",
              "      <td>88.6</td>\n",
              "      <td>98.3</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1560.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>30</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1438 rows Ã— 122 columns</p>\n",
              "</div>"
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "DDVsedvd-UTR"
      },
      "source": [
        "probe_distance = pd.read_table(probe_snp_overlap_fn, usecols=[11, 12]).apply(\n",
        "    lambda x: x.str.split(';')).explode(['probe', 'distance'])\n",
        "bad_probes_dist = probe_distance[probe_distance['distance'].astype('int') < 5].probe\n",
        "\n",
        "probe_chr = pd.read_table(probe_manifest_fn, usecols=[0, 4])\n",
        "bad_probes_chr = probe_chr[(probe_chr['CpG_chrm'] == 'chrM') |\n",
        "                           (probe_chr['CpG_chrm'] == 'chrX') |\n",
        "                           (probe_chr['CpG_chrm'] == 'chrY')].probeID\n",
        "bad_probes = pd.unique(list(bad_probes_chr) + list(bad_probes_dist))"
      ],
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dnxZl3jFC8Jw"
      },
      "source": [
        "skip_f = lambda x: False if x==0 else np.random.choice([True, False], p = [0.8, 0.2])\n",
        "methyl_data = pd.read_table(methyl_fn, index_col='probe_id', skiprows=skip_f).drop(columns = ['#chr', 'start', 'end'])\n",
        "methyl_data = methyl_data.drop(methyl_data.index.intersection(bad_probes)).T\n",
        "methyl_data.index.name = 'NWDID'"
      ],
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "8D9MphDpKa97"
      },
      "source": [
        "metab_data = pd.read_parquet(metab_fn).T\n",
        "protein_data = pd.read_parquet(protein_fn).T"
      ],
      "execution_count": 14,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "DrGiIa-GTq06"
      },
      "source": [
        "exp_data =  pd.read_table(exp_fn, index_col='gene_id').drop(columns = ['#chr', 'start', 'end']).T\n",
        "exp_data.index.name = 'NWDID'"
      ],
      "execution_count": 15,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ADD9wIBmVKE2"
      },
      "source": [
        "common_inds = exp_data.index.intersection(methyl_data.index).intersection(metab_data.index).intersection(protein_data.index).intersection(meta_df.index)\n",
        "\n",
        "exp_data = exp_data.loc[common_inds]\n",
        "methyl_data = methyl_data.loc[common_inds]\n",
        "metab_data = metab_data.loc[common_inds]\n",
        "protein_data = protein_data.loc[common_inds]"
      ],
      "execution_count": 16,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Some inds are still duplicated in the meta_df, and there isn't always a map\n",
        "#   between the sample and the ind in the mol dataset. Thus drop duplicate with\n",
        "#   the most missing data in the phenotype file. This seems to work.\n",
        "meta_df = meta_df.loc[common_inds].reset_index(drop=True).sort_values(\n",
        "    by=['NWDID', 'num_nans']).drop_duplicates(subset='NWDID')"
      ],
      "metadata": {
        "id": "cmRTMINzeinm"
      },
      "execution_count": 17,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "K3jRyZykOY2F"
      },
      "source": [
        "methyl_data = filter_data(methyl_data, max_missing=0)"
      ],
      "execution_count": 18,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "oKhxtZkNYHVb"
      },
      "source": [
        "exp_data.to_csv(exp_data_save_fn)\n",
        "methyl_data.to_csv(methyl_data_save_fn)\n",
        "metab_data.to_csv(metab_data_save_fn)\n",
        "protein_data.to_csv(protein_data_save_fn)\n",
        "meta_df.to_csv(meta_df_save_fn)"
      ],
      "execution_count": 19,
      "outputs": []
    }
  ]
}